### PR TITLE
Refactoring and Introduction of Keycloak as SSO

### PR DESCRIPTION
**Pull request by buehner**
This PR splits the SHOGun into multiple maven submodules and docker:

- `shogun-boot`: Keeps the spring-boot related classes
- `shogun-config`: Keeps the configs
- `shogun-lib`: Keeps the inner shogun logic
- `shogun-gs-interceptor`:  Keeps the geoserver interceptor code

It also introduces KeyCloak as a SSO module.

Pushing on this branch will automatically deploy the beta-snapshot on the nexus!

### CODE REVIEW

File: .github/workflows/four-beta.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+name: Build and publish SHOGun
+
+on:
+  push:
+    branches:
+      - four-beta
```
Comment (ahennr) [inline .github/workflows/four-beta.yml]: Is this workflow needed after merge?

File: pom.xml [hunk 2]
```diff
@@ -10,7 +10,7 @@
 
   <groupId>de.terrestris</groupId>
   <artifactId>shogun</artifactId>
-  <version>3.1.1-SNAPSHOT</version>
+  <version>4.0.0-BETA-SNAPSHOT</version>
```
Comment (ahennr) [inline pom.xml]: `4.0.0-SNAPSHOT` may be more appropriate here?

File: shogun-boot/pom.xml [hunk 3]
```diff
@@ -5,10 +5,9 @@
   <parent>
     <groupId>de.terrestris</groupId>
     <artifactId>shogun</artifactId>
-    <version>3.1.1-SNAPSHOT</version>
+    <version>4.0.0-BETA-SNAPSHOT</version>
```
Comment (ahennr) [inline shogun-boot/pom.xml]: see above comment, maybe maven variable `project.version` can be used here?

File: shogun-boot/src/main/java/de/terrestris/shogun/boot/config/BootWebSecurityConfig.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.web.csrf.CookieCsrfTokenRepository;
+
+@Configuration
+@EnableWebSecurity
+public class BootWebSecurityConfig extends WebSecurityConfig {
+
+//    @Autowired
```
Comment (ahennr) [inline shogun-boot/src/main/java/de/terrestris/shogun/boot/config/BootWebSecurityConfig.java]: Can this be removed?

File: shogun-config/pom.xml [hunk 5]
```diff
@@ -0,0 +1,80 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>de.terrestris</groupId>
+    <artifactId>shogun</artifactId>
+    <version>4.0.0-BETA-SNAPSHOT</version>
```
Comment (ahennr) [inline shogun-config/pom.xml]: See above comment

File: shogun-gs-interceptor/pom.xml [hunk 6]
```diff
@@ -0,0 +1,161 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>de.terrestris</groupId>
+    <artifactId>shogun</artifactId>
+    <version>4.0.0-BETA-SNAPSHOT</version>
```
Comment (ahennr) [inline shogun-gs-interceptor/pom.xml]: See above comment

File: shogun-lib/pom.xml [hunk 7]
```diff
@@ -5,16 +5,14 @@
   <parent>
     <groupId>de.terrestris</groupId>
     <artifactId>shogun</artifactId>
-    <version>3.1.1-SNAPSHOT</version>
+    <version>4.0.0-BETA-SNAPSHOT</version>
```
Comment (ahennr) [inline shogun-lib/pom.xml]: See above comment

**Comment by KaiVolland** [comment]
I'd remove the github workflow and adapt the versions in a small followup if this ok to you?

**Pull request merged**